docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+
+## Primitive Types
+
+Verona treats numeric types specially in the compiler but their representaiton isn't special in the language.
+
+Verona's numeric types are singleton types (no fields) that have their machine representaiton based on their names.
+So, `I32` is a signed integer with 32-bit width (in LLVM's parlour, an `i32`) and so on.
+
+The full list of primitive numeric types are:
+ * Unsigned integers: `U8`, `U16`, `U32`, `U64` and `U128`
+ * Signed integers: `I8`, `I16`, `I32`, `I64` and `I128`
+ * Floating Point: `F32` and `F64`
+ * System types: `ULength` and `ILength` (machine-word sized integers)
+
+Note, integer literals have no defined type and their types (and representation) will be inferred by the compiler depending on the type of the variable being assigned to.
+
+## Type Unions
+
+Type unions means a variable can be either type on the list.
+
+For example:
+```ts
+foobar(a : (A | B | C))
+{
+  ...
+}
+
+main()
+{
+  let a : A;
+  let b : B;
+  let d : D;
+
+  // This is fine
+  foobar(a);
+  foobar(b);
+
+  // This is not
+  foobar(d); // ERROR
+}
+```
+
+The imlementation of `foobar` will have to cope with all three types, either via a `match` statement or taking advantage of some behaviour that is common to all three (say, an interface they all implement).
+
+The compiler is free, however, to optimise the calls by creating two versions of `foobar` and replace the calls to the generic `foobar` for their specific ones.
+If the semantics is defined at compile time, this can simplifies a `match` statement into their specific matches and even inline the call altogether.
+
+Example:
+```ts
+// The generic one, if there are still dynamic cases
+foobar(a : (A | B | C)) { ... }
+
+// The specific specialisations used in defined cases
+foobar(a : A) { ... }
+foobar(b : B) { ... }
+```
+
+## Type Intersections
+
+Intersection means a type must implement all guarantees of all types involved.
+
+For example:
+```ts
+interface Serialisable { toString() : String; };
+
+main()
+{
+  let x : (A & Serialisable & mut) = ...;
+}
+```
+
+In the above example, the type of `x` must conform to both `A` and `Serialisable` guarantees (methods, fields) as well as the `mutable` capability.
+
+Every type in Verona must have a capability, so:
+```ts
+var x : (A & imm); // this is ok
+
+var y : B; // this is not
+```
+
+Type alias can be used to make it easier to use types, for example:
+```ts
+class _String { ... };
+
+// We mostly want strings as values
+type String = (_String & imm);
+
+// But we may want to change it
+type MutableString = (_String & mut);
+```
+
+## Tuples
+
+Tuples allow unnamed types to encompass multiple objects of different types.
+Unlike classes, tuples don't have fields (or has unnamed fields), so you can't access them by name.
+
+Here's a tuple example:
+```ts
+// Declare a tuple type
+var x : (String, U64);
+
+// Tuple literals
+x = ("hi", 42);
+
+// Access elements
+(var s, var i) = x;
```

Comment:
If x's types are `mut`, then I thought `s` and `i` are references to the internal elements. If they're `imm` or a primitive, then it's a copy. If they're `iso`, then it's a move?

This point wasn't well discussed, but we also talked about using Pony's `x._1` as accessor if nothing else works.

Reply:
I don't remember if we actually came up with a way of accessing tuple elements (@sylvanc?).  You can't have an accessor function because the type of the function depends on the value.  If you wanted to use the `apply` sugar then `x(n)` would have to have type `String | U64` (or probably `String | U64 | NotAThing`), so you'd then need to match on the result (which is annoying because for concrete values of `n` you know the type.  C++ uses `get<int>` to force the index to be a compile-time constant, which is quite ugly (and I'm not sure if Verona has enough support in generics to allow this).  As I recall, the Pony-inspired suggestion was to give names to the fields such as `_1`, `_2`, (which is how we got onto the discussion of `_` as a visibility specifier, so maybe `#1`, `#2`, and so on, possibly starting from `#0`).  

If any of the fields are `iso` then decomposition must be a destructive read that destroys the *entire* tuple because otherwise you have an `iso` field with no value, which would introduce UB.  It think it's really confusing is decomposition is sometimes destructive and sometimes not, so I'd be inclined to say that it's always destructive.

Reply:
Do you mean decomposition accessing all fields (like `(a, b) = c`) or also just one (like `a = c#1;`)?

If all, then it's ok for it to be destructive, but if it's also some, then what happens to the rest of the elements?

And if index access isn't destructive, then what happens if the field is an `iso`?

Reply:
`a = c.#1` would have the same semantics for a tuple as `a = obj.field` would for any field.  I don't remember where we landed on `=` but this may give you a `mut` view of the `iso`.

My worry is that structured binding is used for multiple return values.  If I do:

```
example() : (mut & T, iso & X);
...
(var r1, var r2) = example();
```

I want `r2` to give me an `X & iso`.  I would be very confused if it gave me an `N & mut` and no way for me to recover the `iso` (presumably the region would be deallocate at the end of the stack frame and I wouldn't be allowed to store `r2` anywhere on the heap).  This means that I want it to extract the `iso` destructively.  

If the `iso` is extracted destructively, then the field is now invalid and I may not access it again.  I can't destroy a field of an object without destroying the object, so reading an `iso` out via structured binding would destroy the object.

Now imagine I have code like this:

```verona
var tuple : (T & mut, X & imm) = ...;
...

(var a, var b) = tuple;

...

foo = tuple.#1;
```

This is all fine if it's only reading an `iso` that makes this a destructive operation.  But now if I add an `iso` field to tuple, then my later `tuple.#1` becomes invalid *even though the field of the tuple that it references has not changed*.   

If we say that structured binding always destroys the object that it refers to then the above code is consistently disallowed.  I can see a few other valid options here.  For example:

 - Structured binding is not allowed on tuples with `iso` fields (I don't like this one because it eliminates a load of valuable uses for this, but it is a simple rule).
 - Structured binding on `iso` tuples destructively reads all fields, structured binding on `mut` and `imm` tuples does not, structured binding on non-`iso` tuples containing `iso` fields is not permitted.
 - Structured binding gives references to the fields, not copies of their values (I don't think we have a notion of reference-to-field anywhere else in the language though.  You might be able to fudge this with `embed` by making every tuple element an `embed` object with a single field, so you're actually taking a pointer to the object that contains one field but that feels very clunky).

The second of these would cause an error in the above example on the structured binding line, but that would already need modifying because the tuple structure changed.  This could then be fixed by making the tuple `iso`, which causes breakage in the same place but now it's a result of changing the type of the capability of `tuple`, not the type of a field.

Reply:
I think this is an interesting discussion but I don't want to write hypothetical stuff here. Once we define what accessing tuple elements do we can update the document.